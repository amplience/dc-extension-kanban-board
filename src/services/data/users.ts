import type { User } from 'dc-extensions-sdk/dist/types/lib/components/Users';
import type { DcExtensionClient } from '../dc-extension-client';

export async function fetchAll(client: DcExtensionClient): Promise<User[]> {
  return await client.dashboardSdk.users.list();
}
