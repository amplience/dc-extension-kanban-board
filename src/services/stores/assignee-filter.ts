import type { User } from 'dc-extensions-sdk/dist/types/lib/components/Users';
import { persistedWritable } from './persisted-store';

export const assigneesFilter = persistedWritable<User[]>('assigneesFilter', []);
