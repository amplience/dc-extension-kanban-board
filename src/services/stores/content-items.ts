import type { FacetedContentItem } from 'dc-management-sdk-js';
import { writable } from 'svelte/store';

function initContentItems() {
  const { subscribe, set, update } = writable<FacetedContentItem[]>([]);

  let contentItems: FacetedContentItem[] = [];
  subscribe((items) => (contentItems = items));
  return {
    subscribe,
    set,
    update,
    getByStatusId: (id: string) =>
      contentItems.filter((item) => item?.workflow?.state === id),
  };
}

export const contentItems = initContentItems();
