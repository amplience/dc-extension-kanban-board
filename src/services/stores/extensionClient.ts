import { writable } from 'svelte/store';
import { DcExtensionClient } from '../dc-extension-client';

export const extensionClient = writable<DcExtensionClient>(
  new DcExtensionClient()
);
