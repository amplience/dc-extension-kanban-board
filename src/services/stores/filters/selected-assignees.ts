import { persistedWritable } from '../persisted-store';

export const selectedAssignees = persistedWritable<string[]>(
  'selectedAssignees',
  []
);
