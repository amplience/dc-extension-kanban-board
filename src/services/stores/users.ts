import type { User } from 'dc-extensions-sdk/dist/types/lib/components/Users';
import { writable } from 'svelte/store';

export const users = writable<User[]>([]);
