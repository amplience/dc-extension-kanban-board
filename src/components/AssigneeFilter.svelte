<script lang="ts">
  import type { User } from 'dc-extensions-sdk/dist/types/lib/components/Users';
  import { onMount } from 'svelte';
  import FilterIcon from '../assets/icons/ic-filter.svg';
  import { selectedAssignees } from '../services/stores/filters/selected-assignees';
  import { users } from '../services/stores/users';
  import FilterChip from './FilterChip.svelte';
  import Icon from './Icon.svelte';
  import Overlay from './Overlay.svelte';

  let sectionElement: HTMLElement;
  let isModalVisible: boolean = false;
  let modalPositionStyle = '';
  let assignees: User[] = [];

  onMount(() => {
    assignees = $selectedAssignees;
  });

  function updateAssigneeFilter() {
    $selectedAssignees = assignees;
  }

  function removeAssigneeFromFilter(id: string) {
    removeAssignee(id);
    updateAssigneeFilter();
  }

  function updateAssignees(user: User) {
    const existing = assignees.some((assignee) => assignee.id === user.id);

    if (existing) {
      removeAssignee(user.id);
    } else {
      assignees.push(user);
    }

    updateAssigneeFilter();
  }

  function removeAssignee(id: string) {
    assignees = assignees.filter((assignee) => assignee.id !== id);
  }

  function showModal() {
    const targetBound = sectionElement.getBoundingClientRect();
    modalPositionStyle = [
      `top: ${targetBound.y + targetBound.height}px`,
      `left: ${targetBound.x}px`,
    ].join(';');

    isModalVisible = true;
  }

  function onCancelClick() {
    isModalVisible = false;
  }

  function assigneeIncludedInFilter(userId: string) {
    return assignees.some((assignee) => assignee.id === userId);
  }
</script>

<style lang="scss">
  section {
    background-color: #fff;
    padding: 5px;
    position: relative;
  }

  .assignee-filter {
    display: flex;
    flex-direction: row;
    align-items: center;
  }

  .assignee-filter span {
    font-size: 13px;
    font-weight: 400;
    padding-right: 10px;
    cursor: pointer;
  }

  .assignee-filter div.filter-icon {
    height: 20px;
    width: 20px;
    padding: 5px;
    background-color: hsl(0, 0%, 90%);
    border-radius: 4px;
    margin-right: 8px;
    cursor: pointer;
  }

  .assignee-filter div.assignee-names {
    display: flex;
    flex-direction: row;
  }

  .modal-popup {
    background-color: #fff;
    position: fixed;
    width: 500px;
    --webkit-box-shadow: 0 3px 13px rgba(0, 0, 0, 0.2);
    box-shadow: 0 3px 13px rgba(0, 0, 0, 0.2);
    z-index: 3;
  }
</style>

{#if isModalVisible}
  <Overlay onClick={onCancelClick} />
{/if}
<section bind:this={sectionElement}>
  <div class="assignee-filter">
    <span on:click={showModal}>Assignee</span>
    <div class="assignee-names">
      {#if $selectedAssignees.length < 1}
        <FilterChip label="All" on:click={showModal} clickable={true} />
      {:else}
        {#each $selectedAssignees as assignee}
          <FilterChip
            label={`${assignee.firstName} ${assignee.lastName}`}
            removeable={true}
            on:close={() => removeAssigneeFromFilter(assignee.id)}
            on:click={showModal}
            clickable={true} />
        {/each}
      {/if}
    </div>
    <div class="filter-icon" on:click={showModal}>
      <Icon icon={FilterIcon} width="20px" height="20px" />
    </div>
  </div>
  {#if isModalVisible}
    <div class="modal-popup" style={modalPositionStyle}>
      <ul>
        {#each $users as user}
          <li>
            <label for={user.id}>{user.firstName} {user.lastName}</label>
            <input
              type="checkbox"
              value={user.id}
              name={user.id}
              on:change={() => updateAssignees(user)}
              checked={assigneeIncludedInFilter(user.id)} />
          </li>
        {/each}
      </ul>
    </div>
  {/if}
</section>
