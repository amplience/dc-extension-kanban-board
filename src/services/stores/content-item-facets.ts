import type { User } from 'dc-extensions-sdk/dist/types/lib/components/Users';
import type { FacetField } from 'dc-management-sdk-js';
import { derived } from 'svelte/store';
import { selectedAssignees } from './filters/selected-assignees';

export const contentItemFacets = derived(
  [selectedAssignees],
  ([$selectedAssignees]) => {
    const facets: FacetField[] = [];
    if ($selectedAssignees.length > 0) {
      facets.push(createAssigneeFacet($selectedAssignees));
    }
    return facets;
  }
);

function createAssigneeFacet(users: User[]): FacetField {
  return {
    facetAs: 'ENUM',
    field: 'assignees',
    filter: { type: 'IN', values: users.map((user) => user.id) },
    name: 'assignees',
  };
}
