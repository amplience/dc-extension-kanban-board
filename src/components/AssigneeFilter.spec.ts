import { getByTestId, render } from '@testing-library/svelte';
import { tick } from 'svelte';
import { selectedAssignees } from '../services/stores/filters/selected-assignees';
import { users } from '../services/stores/users';
import AssigneeFilter from './AssigneeFilter.svelte';

describe('AssigneeFilter', () => {
  const usersList = [
    {
      id: 'ID_1',
      firstName: 'TEST',
      lastName: 'USER',
      email: 'test-user@bigcontent.io',
    },
    {
      id: 'ID_2',
      firstName: 'ANOTHER',
      lastName: 'USER',
      email: 'another-user@bigcontent.io',
    },
  ];

  beforeEach(() => {
    selectedAssignees.set([]);
    users.set(usersList);
  });

  it('should render the assignee filter', () => {
    const { container } = render(AssigneeFilter);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render the assignee filter with selected assignees', async () => {
    selectedAssignees.set(['ID_1', 'ID_2']);

    const { container } = render(AssigneeFilter);
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render the modal with a list of users', async () => {
    const { container } = render(AssigneeFilter);

    const filterIcon = getByTestId(container, 'filter-icon');
    filterIcon.click();
    await tick();
    expect(container.firstChild).toMatchSnapshot();
  });

  it('should render the modal with a list of users with checkboxes ticked', async () => {
    selectedAssignees.set(['ID_1', 'ID_2']);
    const { container } = render(AssigneeFilter);

    const filterIcon = getByTestId(container, 'filter-icon');
    filterIcon.click();
    await tick();
    expect(container.firstChild).toMatchSnapshot();
  });
});
