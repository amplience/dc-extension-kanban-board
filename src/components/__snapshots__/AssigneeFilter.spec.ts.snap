// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AssigneeFilter should render the assignee filter 1`] = `
<div>
   
  <section
    class="svelte-1ggh271"
  >
    <div
      class="assignee-filter svelte-1ggh271"
    >
      <div
        class="filter-icon  svelte-1ggh271"
        data-testid="filter-icon"
      >
        <div
          class=" svelte-zpbvvv"
          style="--icon-width: 20px; --icon-height: 20px;"
        >
          <object
            src="ic-filter.svg/"
          />
        </div>
      </div>
       
      <span
        class="svelte-1ggh271"
      >
        Assignee
      </span>
       
      <div
        class="assignee-names svelte-1ggh271"
      >
        <div
          class=" clickable  svelte-h28tab"
          data-testid="chip"
        >
          <span
            class="label svelte-h28tab"
          >
            All
          </span>
           
        </div>
      </div>
    </div>
     
  </section>
</div>
`;

exports[`AssigneeFilter should render the assignee filter with selected assignees 1`] = `
<div>
   
  <section
    class="svelte-1ggh271"
  >
    <div
      class="assignee-filter svelte-1ggh271"
    >
      <div
        class="filter-icon has-assignees svelte-1ggh271"
        data-testid="filter-icon"
      >
        <div
          class=" svelte-zpbvvv"
          style="--icon-width: 20px; --icon-height: 20px;"
        >
          <object
            src="ic-filter.svg/"
          />
        </div>
      </div>
       
      <span
        class="svelte-1ggh271"
      >
        Assignee
      </span>
       
      <div
        class="assignee-names svelte-1ggh271"
      >
        <div
          class=" clickable removeable svelte-h28tab"
          data-testid="chip"
        >
          <span
            class="label svelte-h28tab"
          >
            TEST USER
          </span>
           
          <span
            class="cross svelte-h28tab"
            data-testid="remove-chip-button"
          >
            <div
              class=" svelte-zpbvvv"
              style="--icon-width: 8px; --icon-height: 8px;"
            >
              <object
                src="ic-delete.svg/"
              />
            </div>
          </span>
        </div>
        <div
          class=" clickable removeable svelte-h28tab"
          data-testid="chip"
        >
          <span
            class="label svelte-h28tab"
          >
            ANOTHER USER
          </span>
           
          <span
            class="cross svelte-h28tab"
            data-testid="remove-chip-button"
          >
            <div
              class=" svelte-zpbvvv"
              style="--icon-width: 8px; --icon-height: 8px;"
            >
              <object
                src="ic-delete.svg/"
              />
            </div>
          </span>
        </div>
        
      </div>
    </div>
     
  </section>
</div>
`;

exports[`AssigneeFilter should render the modal with a list of users 1`] = `
<div>
  <div
    class="svelte-1k192f9"
    data-testid="overlay"
    style="z-index: 2"
  />
   
  <section
    class="svelte-1ggh271"
  >
    <div
      class="assignee-filter svelte-1ggh271"
    >
      <div
        class="filter-icon  svelte-1ggh271"
        data-testid="filter-icon"
      >
        <div
          class=" svelte-zpbvvv"
          style="--icon-width: 20px; --icon-height: 20px;"
        >
          <object
            src="ic-filter.svg/"
          />
        </div>
      </div>
       
      <span
        class="svelte-1ggh271"
      >
        Assignee
      </span>
       
      <div
        class="assignee-names svelte-1ggh271"
      >
        <div
          class=" clickable  svelte-h28tab"
          data-testid="chip"
        >
          <span
            class="label svelte-h28tab"
          >
            All
          </span>
           
        </div>
      </div>
    </div>
     
    <div
      class="modal-popup svelte-1ggh271"
      style="top: NaNpx;left: undefinedpx"
    >
      <h3
        class="svelte-1ggh271"
      >
        Assignee
      </h3>
       
      <ul
        class="
      mdc-list
      
      
      
      
      
      
    "
        role="group"
      >
        <li
          aria-checked="false"
          aria-disabled="false"
          class="
    mdc-list-item
    
    
    
    
    
  "
          role="checkbox"
          tabindex="-1"
        >
          <span
            class="mdc-list-item__graphic "
          >
            <div
              class="mdc-checkbox mdc-checkbox--upgraded"
            >
              <input
                class="mdc-checkbox__native-control "
                id="SMUI-form-field-list-0"
                tabindex="-1"
                type="checkbox"
                value="ID_1"
              />
               
              <div
                class="mdc-checkbox__background"
              >
                <svg
                  class="mdc-checkbox__checkmark"
                  viewBox="0 0 24 24"
                >
                  <path
                    class="mdc-checkbox__checkmark-path"
                    d="M1.73,12.91 8.1,19.28 22.79,4.59"
                    fill="none"
                  />
                </svg>
                 
                <div
                  class="mdc-checkbox__mixedmark"
                />
              </div>
               
              <div
                class="mdc-checkbox__ripple"
              />
            </div>
          </span>
          
           
          <label
            class="mdc-list-item__text "
            for="SMUI-form-field-list-0"
          >
            TEST
             
            USER
          </label>
           
        </li>
        
        <li
          aria-checked="false"
          aria-disabled="false"
          class="
    mdc-list-item
    
    
    
    
    
  "
          role="checkbox"
          tabindex="-1"
        >
          <span
            class="mdc-list-item__graphic "
          >
            <div
              class="mdc-checkbox mdc-checkbox--upgraded"
            >
              <input
                class="mdc-checkbox__native-control "
                id="SMUI-form-field-list-1"
                tabindex="-1"
                type="checkbox"
                value="ID_2"
              />
               
              <div
                class="mdc-checkbox__background"
              >
                <svg
                  class="mdc-checkbox__checkmark"
                  viewBox="0 0 24 24"
                >
                  <path
                    class="mdc-checkbox__checkmark-path"
                    d="M1.73,12.91 8.1,19.28 22.79,4.59"
                    fill="none"
                  />
                </svg>
                 
                <div
                  class="mdc-checkbox__mixedmark"
                />
              </div>
               
              <div
                class="mdc-checkbox__ripple"
              />
            </div>
          </span>
          
           
          <label
            class="mdc-list-item__text "
            for="SMUI-form-field-list-1"
          >
            ANOTHER
             
            USER
          </label>
           
        </li>
        
        
      </ul>
      
    </div>
  </section>
</div>
`;

exports[`AssigneeFilter should render the modal with a list of users with checkboxes ticked 1`] = `
<div>
  <div
    class="svelte-1k192f9"
    data-testid="overlay"
    style="z-index: 2"
  />
   
  <section
    class="svelte-1ggh271"
  >
    <div
      class="assignee-filter svelte-1ggh271"
    >
      <div
        class="filter-icon has-assignees svelte-1ggh271"
        data-testid="filter-icon"
      >
        <div
          class=" svelte-zpbvvv"
          style="--icon-width: 20px; --icon-height: 20px;"
        >
          <object
            src="ic-filter.svg/"
          />
        </div>
      </div>
       
      <span
        class="svelte-1ggh271"
      >
        Assignee
      </span>
       
      <div
        class="assignee-names svelte-1ggh271"
      >
        <div
          class=" clickable removeable svelte-h28tab"
          data-testid="chip"
        >
          <span
            class="label svelte-h28tab"
          >
            TEST USER
          </span>
           
          <span
            class="cross svelte-h28tab"
            data-testid="remove-chip-button"
          >
            <div
              class=" svelte-zpbvvv"
              style="--icon-width: 8px; --icon-height: 8px;"
            >
              <object
                src="ic-delete.svg/"
              />
            </div>
          </span>
        </div>
        <div
          class=" clickable removeable svelte-h28tab"
          data-testid="chip"
        >
          <span
            class="label svelte-h28tab"
          >
            ANOTHER USER
          </span>
           
          <span
            class="cross svelte-h28tab"
            data-testid="remove-chip-button"
          >
            <div
              class=" svelte-zpbvvv"
              style="--icon-width: 8px; --icon-height: 8px;"
            >
              <object
                src="ic-delete.svg/"
              />
            </div>
          </span>
        </div>
        
      </div>
    </div>
     
    <div
      class="modal-popup svelte-1ggh271"
      style="top: NaNpx;left: undefinedpx"
    >
      <h3
        class="svelte-1ggh271"
      >
        Assignee
      </h3>
       
      <ul
        class="
      mdc-list
      
      
      
      
      
      
    "
        role="group"
      >
        <li
          aria-checked="true"
          aria-disabled="false"
          class="
    mdc-list-item
    
    
    
    
    
  "
          role="checkbox"
          tabindex="0"
        >
          <span
            class="mdc-list-item__graphic "
          >
            <div
              class="mdc-checkbox mdc-checkbox--upgraded"
            >
              <input
                class="mdc-checkbox__native-control "
                id="SMUI-form-field-list-2"
                tabindex="-1"
                type="checkbox"
                value="ID_1"
              />
               
              <div
                class="mdc-checkbox__background"
              >
                <svg
                  class="mdc-checkbox__checkmark"
                  viewBox="0 0 24 24"
                >
                  <path
                    class="mdc-checkbox__checkmark-path"
                    d="M1.73,12.91 8.1,19.28 22.79,4.59"
                    fill="none"
                  />
                </svg>
                 
                <div
                  class="mdc-checkbox__mixedmark"
                />
              </div>
               
              <div
                class="mdc-checkbox__ripple"
              />
            </div>
          </span>
          
           
          <label
            class="mdc-list-item__text "
            for="SMUI-form-field-list-2"
          >
            TEST
             
            USER
          </label>
           
        </li>
        
        <li
          aria-checked="true"
          aria-disabled="false"
          class="
    mdc-list-item
    
    
    
    
    
  "
          role="checkbox"
          tabindex="0"
        >
          <span
            class="mdc-list-item__graphic "
          >
            <div
              class="mdc-checkbox mdc-checkbox--upgraded"
            >
              <input
                class="mdc-checkbox__native-control "
                id="SMUI-form-field-list-3"
                tabindex="-1"
                type="checkbox"
                value="ID_2"
              />
               
              <div
                class="mdc-checkbox__background"
              >
                <svg
                  class="mdc-checkbox__checkmark"
                  viewBox="0 0 24 24"
                >
                  <path
                    class="mdc-checkbox__checkmark-path"
                    d="M1.73,12.91 8.1,19.28 22.79,4.59"
                    fill="none"
                  />
                </svg>
                 
                <div
                  class="mdc-checkbox__mixedmark"
                />
              </div>
               
              <div
                class="mdc-checkbox__ripple"
              />
            </div>
          </span>
          
           
          <label
            class="mdc-list-item__text "
            for="SMUI-form-field-list-3"
          >
            ANOTHER
             
            USER
          </label>
           
        </li>
        
        
      </ul>
      
    </div>
  </section>
</div>
`;
